rfcs/0110-inherit-as-list.md

```diff
@@ -0,0 +1,125 @@
+---
+feature: inherit-as-list
+start-date: 2021-10-17
+author: Ryan Burns (@r-burns)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a new Nix syntax `inherit (<attrset>) [ <attrnames> ]`, which
+constructs a list from the values of an attrset.
+
+The goal is to provide a similarly-terse but more principled alternative
+to the often-used `with <attrset>; [ <attrnames> ]`.
+
+# Motivation
+[motivation]: #motivation
+
+It is currently cumbersome to create a list from the values of an attrset.
+If one has an attrset `attrs` and wishes to create a list containing some of
+its values, one could naively write:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+```
+
+To avoid typing `attrs` many times, one will typically use `with` instead:
+
+```nix
+with attrs; [ a b c ]
+```
+
+However, the `with` expression has many well-known drawbacks, such as
+unintuitive shadowing behavior [1][2], prevention of static scope checking [3][4],
+and reduced evaluation performance [3].
+
+* [1] https://github.com/NixOS/nix/issues/490
+* [2] https://github.com/NixOS/nix/issues/1361
+* [3] https://github.com/NixOS/nixpkgs/pull/101139
+* [4] https://nix.dev/anti-patterns/language#with-attrset-expression
+
+Nonetheless, Nix expression authors are subtly guided toward the `with` form
+because it is (or at least appears) simpler than any existing alternatives.
+Some alternatives are suggested in
+https://nix.dev/anti-patterns/language#with-attrset-expression, but as these
+are more verbose and complex than `with`, they are rarely used.
+
+The goal of this RFC is to provide a similarly-terse alternative which avoids
+these drawbacks.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed expression syntax is:
+
+```nix
+inherit (attrs) [ a b c ]
+```
+
+The `inherit` keyword is intentionally reused so as to not add any new
```

Comment:
It might be confusing that for attribute sets, the `inherit` keyword is on the inside but here it is on the outside.

Having

```nix
[
inherit (attrs) a b c;
some
other
elements
]
```

would be slightly more predictable and maybe even more convenient (no need to concatenate multiple arrays like with `with`) but probably even worse for parsing.

Reply:
Yes, I considered that asymmetry but stuck with the current approach for the initial draft of this RFC - the one I have right now is dead simple to implement for an MVP. I like your example, but it might require a complicated grammar.

Another alternative would be to reuse a different keyword so as to avoid the false parallel with the `inherit (<attrset>) <attrnames>;` form.